planner/core/planbuilder.go

```diff
@@ -1115,35 +1115,60 @@ func (b *PlanBuilder) buildShow(show *ast.ShowStmt) (Plan, error) {
 	for _, col := range p.schema.Columns {
 		col.UniqueID = b.ctx.GetSessionVars().AllocPlanColumnID()
 	}
-	mockTablePlan := LogicalTableDual{}.Init(b.ctx)
+	mockTablePlan := LogicalTableDual{placeHolder: true}.Init(b.ctx)
 	mockTablePlan.SetSchema(p.schema)
+	var err error
+	var np LogicalPlan
+	np = mockTablePlan
 	if show.Pattern != nil {
 		show.Pattern.Expr = &ast.ColumnNameExpr{
 			Name: &ast.ColumnName{Name: p.Schema().Columns[0].ColName},
 		}
-		expr, _, err := b.rewrite(show.Pattern, mockTablePlan, nil, false)
+		np, err = b.buildSelection(np, show.Pattern, nil)
 		if err != nil {
 			return nil, err
 		}
-		p.Conditions = append(p.Conditions, expr)
 	}
 	if show.Where != nil {
-		conds := splitWhere(show.Where)
-		for _, cond := range conds {
-			expr, _, err := b.rewrite(cond, mockTablePlan, nil, false)
-			if err != nil {
-				return nil, err
-			}
-			p.Conditions = append(p.Conditions, expr)
+		np, err = b.buildSelection(np, show.Where, nil)
+		if err != nil {
+			return nil, err
 		}
-		err := p.ResolveIndices()
+	}
+	if np != mockTablePlan {
+		fieldsLen := len(mockTablePlan.schema.Columns)
+		proj := LogicalProjection{Exprs: make([]expression.Expression, 0, fieldsLen)}.Init(b.ctx)
+		schema := expression.NewSchema(make([]*expression.Column, 0, fieldsLen)...)
+		for _, col := range mockTablePlan.schema.Columns {
+			proj.Exprs = append(proj.Exprs, col)
+			newCol := col.Clone().(*expression.Column)
+			newCol.UniqueID = b.ctx.GetSessionVars().AllocPlanColumnID()
+			schema.Append(newCol)
+		}
+		proj.SetSchema(schema)
+		proj.SetChildren(np)
+		physical, err := DoOptimize(b.optFlag|flagEliminateProjection, proj)
 		if err != nil {
 			return nil, err
 		}
+		return substitutePlaceHolderDual(physical, p), nil
 	}
 	return p, nil
 }
 
+func substitutePlaceHolderDual(src PhysicalPlan, dst PhysicalPlan) PhysicalPlan {
+	if dual, ok := src.(*PhysicalTableDual); ok && dual.placeHolder {
+		return dst
+	}
+	newChildren := make([]PhysicalPlan, 0, len(src.Children()))
+	for _, child := range src.Children() {
+		newChild := substitutePlaceHolderDual(child, dst)
```

Comment:
how about directly changing the i-th child instead of creating another children slice? For example:
```go
for i, child := range src.Children() {
	src.Children()[i] = substitutePlaceHolderDual(child, dst)
}
```

Reply:
Well, this method is somehow a little tricky. It depends on that the returned children slice by `Children()` is directly referenced to the memory storing the children information. If the slice returned by `Children()` is a copy of that information, then changing the content of that slice will not affect the real children info.

A better way is to provide a `SetChild(i int, newChild PhysicalPlan)` for interface `PhysicalPlan`

Reply:
Updated, PTAL